doc/designs/quic-design/tx-packetiser.md

```diff
@@ -0,0 +1,330 @@
+| I | Valid in Initial packets|
+| H | Valid in Handshake packets|
+| 0 | Valid in 0-RTT packets|
+| 1 | Valid in 1-RTT packets|
+
+_Spec_ is defined as:
+
+| Spec | Description |
+| :---: | --- |
+| N | Not ack-eliciting. |
+| C | does not count toward bytes in flight for congestion control purposes. |
+| P | Can be used to probe new network paths during connection migration. |
+| F | The contents of frames with this marking are flow controlled. |
+
+For `C`, `N` and `P`, the entire packet much consist of only frames with the
+marking for the packet to quality for it.  For example, a packet with an ACK
+frame and a _stream_ frame would qualify for neither the `C` or `N` markings.
+
+### Notes
+
+- Will likely need wrapper structure around the individial frames.
+
+### Packets
+
+Frames are coalesced into packets which are then sent by the record layer.
+The `packet_header` is a pointer to the leading bytes of the packet.
+The `frames` are pointers to the individual frames that make up the
+packet's body.
+It is expected that the record layer will encrypt from the `packet_header` and
+`frames` directly without a copy.
+
+```c
+enum packet_validity_e {
+    QUIC_PACKET_INITIAL,
+    QUIC_PACKET_HANDSHAKE,
+    QUIC_PACKET_0_RTT,
+    QUIC_PACKET_1_RTT
+};
+
+typedef enum packet_validity_e PACKET_VALIDITY;
+
+struct ossl_quic_packet_st {
+    SSL *connection;
+    unsigned char *packet_header;
+    size_t packet_header_length;
+    STACK_OF(OSSL_QUIC_FRAME) *frames;
+
+    QUIC_PN packet_number; /* RFC 9000 12.3 */
+    size_t packet_length;
+
+    /* Pkts - might be better as one field as per ACK Manager */
+    unsigned int initial : 1;
+    unsigned int handshake : 1;
+    unsigned int rtt0 : 1;
+    unsigned int rtt1 : 1;
+    /* Pkts options */
+    PACKET_VALIDITY validity;
+
+    /* Spec */
+    unsigned int no_ack : 1;
+    unsigned int no_congestion_control : 1;
+    unsigned int probing : 1;
+    unsigned int flow_controlled : 1;
+};
+
+typedef struct ossl_quic_packet_st OSSL_QUIC_PACKET;
+```
+
+#### Notes
+
+- `QUIC_PN` defined as uint64_t in ACK Manager.
+- The `QUIC_ACKM_TX_PKT` struture from the ACK Manager overlaps this
+  significantly and the two could be merged.
+- Do we need the distinction between 0-rtt and 1-rtt when both are in
+  the Application Data number space?
+- the `packet_header` will become a structure / union most likely.
+
+Interactions
+------------
+
+The packetiser needs to interact with other modules.  This defines the APIs
+by which it does so.
+
+### Stream Send Buffers
+
+Frames are passed to the packetiser on a per stream basis.
+The frames must be fully formed.  By passing a frame to this function,
+ownership is passed to the packetiser which queues the frames for later
+sending by the record layer.
+
+```c
+__owur int ossl_quic_packetise(SSL_QUIC_TX_PACKETISER *tx, SSL *stream,
+                               const OSSL_QUIC_FRAME *frame,
+                               size_t frame_length);
```

Comment:
I expected the packetiser to determine the length of application data frames - not be told by something external what that length is. Presumably the packetiser knows the maximum size that the packet can be, and we also (presumably) want to send maximum sized packets where ever possible. The packetiser will have to take the data input from various sources and form the complete packet from it. For example let say that the packetiser decides to send some ack frames and a crypto frame, and its got 1k of "space" left in the packet which it can fill with application data. It needs to take that much data from the steam buffers. Surely, only the packetiser knows the amount of space available and can form the application data frames accordingly?

Reply:
The design specifies that complete frames are passed into the packetiser.  Either we need to change this or the  stream send buffers need to know the length (which means the packetiser should have a call to make this available).

I'm imagining that several packets will be on the go at any one time.
Packetising is equivalent to the bin packing problem, so this should ease things.  Also, there are frames that are best sent separately (e.g. ack frames don't need an ack if sent without any frames that do need an ack, likewise they don't count for bytes in flight alone).

Reply:
I was envisaging that the packetiser would have access to a set of stream send buffers. Each stream buffer would know how much data it has available to send - and the packetiser can query that. The packetiser would query each stream buffer in turn and fill the "current" packet with data from each buffer. We will have to have some algorithm to determine how much to take from each one - I can think of a few different strategies, e.g. 2 possibilities might be:

1) Take as much data as possible from the first stream buffer, until either we've run out of data or the packet is full. If the packet is not yet full move onto the next stream buffer and repeat. Next time we create a packet continue where we left off, and iterate through each stream buffer in a round-robin fashion.

or

2) Work out how many stream buffers we have. Divide the amount of space available by the amount of stream buffers and fill the packet with a small amount of data from every buffer. If any one stream buffer has less data than its allowance, then increase the allowance of all the other stream buffers accordingly.

Possibly different stream buffers have different priorities. You could envisage a "very high priority" that means "just send the data from this stream buffer as soon as possible and don't mix it with other stream data", which could be used for the ack frames scenario you describe. Or something like that anyway.

Reply:
IMO the best strategy (to avoid stalling multiple streams when packets are lost) is to fill in a packet with as much data available from a single stream as possible. Then in another packet continue with next stream to provide fairness to streams and so on. Of course if a stream does not have enough data to fill a full packet, add data from another stream. So this is something in-between 1 and 2.

Reply:
At least in the MVP I would not try to add priority to streams as prioritization can be done by the application itself by blocking adding more stream data to lower priority streams until it has sent all the high priority stream data.

Reply:
Retries should be done at higher priority.  We could ignore this for MVP but probably shouldn't.  This implies some form of priority system.  A rather difficult one since we also want fairness across all streams and performance.  These latter two could be done after MVP IMO.

Reply:
I think we need to nail done what is doing the frame construction.

The design says the TXP receives frames from several places elsewhere.  Here, it seems the TXP is creating application frames and getting other frames from elsewhere (or should the TXP create all frames?).  We could also add a _framer_ module.

Reply:
I thought the TXP creates the frames (i.e., the framer is part of the TXP). The calling modules just ask it to wrap their content into appropriate frames. I.e., ACK manager asks it to create ACK frames, the TLS handshake RL asks it to create CRYPTO frames, and so on. The TXP then buffers/sends the frames as appropriate.

Reply:
> Retries should be done at higher priority.

For lost frames? Well they just need to be sent before sending any other data for the same stream IMO it is not that complicated.

Reply:
The design speaks otherwise:

> TX Packetizer
> -------------
> 
> This module creates frames from the application data obtained from
> the application. It also receives CRYPTO frames from the TLS Handshake
> Record Layer and ACK frames from the ACK Handling And Loss Detector
> subsystem.

Which means the TXP should be creating the application data frames.

Reply:
Yeah, the design is vague in this regard. Perhaps it should be updated? I.e. `It also receives data for CRYPTO frames from the TLS Handshake Record Layer and data for ACK frames from the ACK Handling And Loss Detector subsystem.` ? I do not know what makes better sense.

Reply:
@mattcaswell said:
> I was envisaging that the packetiser would have access to a set of stream send buffers. Each stream buffer would know how much data it has available to send - and the packetiser can query that. The packetiser would query each stream buffer in turn and fill the "current" packet with data from each buffer.

So I wonder, who would call the appropriate function to get the packetiser to start working?

Me, I was always thinking that it would be other parts that would essentially call the packetiser function to say "hey, I have something for you"...  and while doing so, they could as well include "and here's my data buffer".  As far as I understand, that's more or less exactly how this function should be called?  I'm not sure I understand why the packetiser should do the extra churn of running around all the SSL objects to ask "hey, do you have something for me?"  in this case.

So, in the scenario where the packetiser just gets the signal to start doing something and go asking all the SSL objects it can reach, how does that process get started?